---
title: "rnaseqGene Notebook"
output: html_notebook
---
# rnaseqGene workflow notebook


For this script to function properly, install the rnaseqGene package from bioconductor
```{r dont_run}
# For this script to function properly, install the rnaseqGene package from bioconductor using the commands directly below:
# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
# BiocManager::install("rnaseqGene", version = "3.8")
```


### Section 2.3: locating alignment files
```{r}
library(airway)
indir <- system.file("extdata", package="airway", mustWork=TRUE) # indir variable is just the name of a folder
indir # see?
list.files(indir) # shows us the files in that folder

# importing the sample table into R
csvfile <- file.path(indir, "sample_table.csv")
sampleTable <- read.csv(csvfile, row.names = 1)
sampleTable
```

### Section 2.4: DESeq2 import functions
First, we need to import in the BAM files for our samples
```{r}
# import counts from BAM files
filenames <- file.path(indir, paste0(sampleTable$Run, "_subset.bam"))
file.exists(filenames)

library("Rsamtools")
bamfiles <- BamFileList(filenames, yieldSize=2000000) # only import 2M seqs at a time

seqinfo(bamfiles[1])
```

### Section 2.5: Defining gene models
To extract/summarize count info from the BAM files, we need annotations of where the genes and exons are, namely in GTF format.
```{r}
library("GenomicFeatures")
gtffile <- file.path(indir,"Homo_sapiens.GRCh37.75_subset.gtf")
txdb <- makeTxDbFromGFF(gtffile, format = "gtf", circ_seqs = character())
txdb #transcript db

ebg <- exonsBy(txdb, by="gene")
ebg # "exons by gene"
```

### Section 2.6: Read Counting & making a Summarized Experiment S4 object
```{r}
library("GenomicAlignments")
library("BiocParallel")
register(SerialParam()) # apparently, this drops us back down to 1 core use for the demo
# 
# create the `se` summarized experiment object
se <- summarizeOverlaps(features=ebg, reads=bamfiles,
                        mode="Union",
                        singleEnd=FALSE,
                        ignore.strand=TRUE,
                        fragments=TRUE )
se
```

### Section 2.7: about the SummarizedExperiment object
```{r}
se

dim(se)
assayNames(se)
head(assay(se), 3)
colSums(assay(se))
rowRanges(se)
str(metadata(rowRanges(se))) # str is short for "internal STRucture"

### to view colData
colData(se) # no columns!
### let's properly fill this slot
colData(se) <- DataFrame(sampleTable)
colData(se)
```

## Section 3: The DESeqDataSet object, sample information and the design formula
```{r}
se$cell
se$dex

library("magrittr")
se$dex %<>% relevel("untrt")
se$dex
# this funny pipe that sends the changes back to the original object is equivalent to this:
# se$dex <- relevel(se$dex, "untrt")

```

### Section 3.1: Starting from a SummarizedExperiment

(yes, it over-writes our old `se` object. I don't agree with this choice of just erasing the work we did before, but I'll live with it.

```{r}
data("airway")
se <- airway

se$dex %<>% relevel("untrt")
se$dex

round( colSums(assay(se)) / 1e6, 1 )
colData(se)

# only putting this here because i've organized my chunks by section headers
library("DESeq2")
dds <- DESeqDataSet(se, design = ~ cell + dex)
```


### Section 3.2 <skip> Starting from count matrices

We're going to skip 3.2 in class because it covers very similar steps to the first way we imported in data, and the main differences are that instead of importing from the BAMs, we would be running an outside counting tool that gave us a csv or tsv of counts, and building our S4 objects from that entry point.

## Section 4: Exploratory analysis and visualization

filtering, modeling variance examples, PCAs

### Section 4.1: Pre-filtering the dataset
```{r}
nrow(dds)
dds <- dds[ rowSums(counts(dds)) > 1, ]
nrow(dds)

```



nrow(dds)
[compare the presented filter with the one used in the BAD days Day 3 workflow](https://bitsandchips.me/BAD_days/RNA-sequencing.html#data-pre-processing)

### Section 4.2: The variance stabilizing transformation and the rlog
```{r}
lambda <- 10^seq(from = -1, to = 2, length = 1000)
cts <- matrix(rpois(1000*100, lambda), ncol = 100)
library("vsn")
# this install did not come with the workflow, but i bet they thought it would VVVVVV
# install.packages("hexbin")
library(hexbin)
meanSdPlot(cts, ranks = FALSE)

log.cts.one <- log2(cts + 1)
meanSdPlot(log.cts.one, ranks = FALSE)

vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)
colData(vsd)

rld <- rlog(dds, blind = FALSE)
head(assay(rld), 3)
```

### Section 4.3: Sample distances
```{r}

```
### Section 4.4: PCA plot
```{r}

```
### Section 4.5: MDS plot
_omitting due to time & rough similarity to PCA_

## Section 5: Differential expression analysis
### Section 5.1: Running the differential expression pipeline
```{r}

```

### Section 5.2: Building the results table
```{r}

```
### Section 5.3: Other comparisons
_omitting for time_
### Section 5.4: Multiple testing
```{r}

```

## Section 6: Plotting results
### Section 6.1: Counts plot
```{r}

```
### Section 6.2: MA-plot
```{r}

```

### Section 6.3: Gene clustering
heatmaps
```{r}

```

### Section 6.4: Independent filtering
_omitting for time_
## Section 7: Annotating and exporting 
Adding HGNC gene symbols and Entrez IDs to our results (res)
```{r}

```

### Section 7.1: Exporting results
Convert `res` to a data frame and write to csv
```{r}

```

### Section 7.2: Plotting fold changes in genomic space
### Section 8.2: Removing hidden batch effects
### Section 8.1: Using SVA with DESeq2
### Section 8.2: Using RUV with DESeq2